pkg/operator/kubeletversionskewcontroller/kubelet_version_skew_controller.go

```diff
@@ -0,0 +1,163 @@
+	KubeletVersionSkewAtLimitReason       = "KubeletVersionSkewAtLimit"
+	KubeletVersionUnsupportedReason       = "KubeletVersionUnsupported"
+)
+
+// KubeletVersionSkewController sets Upgradeable=False if the kubelet
+// version on a node prevents upgrading to a supported OCP version.
+//
+// If OCP minor version is odd, kubelet versions older than 1 minor
+// version of the API server version set Upgradeable=False.
+//
+// If OCP minor version is even, kubelet versions must be in sync with
+// API server version in order to set Upgradeable=True.
+type KubeletVersionSkewController interface {
+	factory.Controller
+}
+
+func NewKubeletVersionSkewController(
+	operatorClient v1helpers.OperatorClient,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	recorder events.Recorder,
+) *kubeletVersionSkewController {
+	c := &kubeletVersionSkewController{
+		operatorClient:     operatorClient,
+		nodeLister:         kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Lister(),
+		apiServerVersion:   semver.MustParse(status.VersionForOperandFromEnv()),
+		minUpgradeableSkew: minUpgradeableKubeletSkewForOCPVersion(status.VersionForOperatorFromEnv()),
+	}
+	c.Controller = factory.New().
+		WithSync(c.sync).
+		WithInformers(kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Informer()).
+		ToController("KubeletVersionSkewController", recorder.WithComponentSuffix("kubelet-version-skew-controller"))
+	return c
+}
+
+func minUpgradeableKubeletSkewForOCPVersion(v string) int {
+	switch semver.MustParse(v).Minor % 2 {
+	case 0: // even OCP versions
+		return 0
+	case 1: // odd OCP versions
+		return -1
+	default:
+		panic("should not happen")
+	}
+}
+
+type kubeletVersionSkewController struct {
+	factory.Controller
+	operatorClient     v1helpers.OperatorClient
+	nodeLister         corev1listers.NodeLister
+	apiServerVersion   semver.Version
+	minUpgradeableSkew int
+}
+
+func (c *kubeletVersionSkewController) sync(_ context.Context, _ factory.SyncContext) error {
+	operatorSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if !management.IsOperatorManaged(operatorSpec.ManagementState) {
+		return nil
+	}
+
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	var errorMsgs []string
+	var skewedToTheLimitMsgs []string
+	var skewedButOKMsgs []string
+	var syncedMsgs []string
+	var unsupportedMsgs []string
+
+	// for each node, check kubelet version
+	for _, node := range nodes {
+		kubeletVersion, err := nodeKubeletVersion(node)
+		if err != nil {
+			errorMsgs = append(errorMsgs, fmt.Sprintf("Unable to determine kubelet version on node %s: %s", node.Name, err))
+			continue
+		}
+		skew := int(kubeletVersion.Minor - c.apiServerVersion.Minor)
+		switch {
+		case skew == 0: // synced
+			syncedMsgs = append(syncedMsgs, fmt.Sprintf("Kubelet version on node %s is in synced with the expected API server version.", node.Name))
```

Comment:
probably point out here (and possibly also in the `KubeletVersionSynced` reason) that "synced" means major/minor match, but does not imply a patch match.  Because we don't want to give folks with 1.21.1 kubelets around the false impression that they are completely caught up with a cluster that would prefer 1.21.99 kubelets.  The APIs will be synchronized, but the implementations will differ in bug exposure.  Maybe `KubeletAPIVersionSynced` or some such might be enough to get the focus onto the API and off of the implementations?  Or something else?

Reply:
actually, probably point this out in the reason and your `Kubelet and API server versions are synced` string.  I don't think we need `syncedMsgs` at all, because the `KubeletVersionSyncedReason` case is the `default` below.  This assumes you buy [my message-denoising argument][1].

[1]: https://github.com/openshift/cluster-kube-apiserver-operator/pull/1199#discussion_r690794774

Reply:
> probably point out here (and possibly also in the `KubeletVersionSynced` reason) that "synced" means major/minor match, but does not imply a patch match.
 
I agree, I'll re-work the reasons/messages to make that clear.